loadbalancer/algorithm_test.go

```diff
@@ -312,6 +312,46 @@ func TestConsistentHashSearch(t *testing.T) {
 	}
 }
 
+func TestConsistentHashBoundedLoadSearch(t *testing.T) {
+	endpoints := []string{"http://127.0.0.1:8080", "http://127.0.0.2:8080", "http://127.0.0.3:8080"}
+	r, _ := http.NewRequest("GET", "http://127.0.0.1:1234/foo", nil)
+	route := NewAlgorithmProvider().Do([]*routing.Route{{
+		Route: eskip.Route{
+			BackendType: eskip.LBBackend,
+			LBAlgorithm: ConsistentHash.String(),
+			LBEndpoints: endpoints,
+		},
+	}})[0]
+	ch := route.LBAlgorithm.(consistentHash)
+	ctx := &routing.LBContext{Request: r, Route: route, Params: map[string]interface{}{ConsistentHashBalanceFactor: 1.25}}
+	noLoad := ch.Apply(ctx)
+	nonBounded := ch.Apply(&routing.LBContext{Request: r, Route: route, Params: map[string]interface{}{}})
+
+	if noLoad != nonBounded {
+		t.Error("When no endpoints are overloaded, the chosen endpoint should be the same as standard consistentHash")
+	}
+	// now we know that noLoad is the endpoint which should be requested for somekey if load is not an issue.
+	addInflightRequests(noLoad, 20)
```

Comment:
After some thought I think we have a problem here. It could not happen that single backend gets +20 inflight requests when requests are balanced by the algorithm because algorithm would prevent them from going to the same backend (avg load = 20/3 = 6.6, max allowed load is 1.25*6.6=8.33 which is less than 20)

I think we need to change test approach and e.g. issue N requests (with a same key) and check that
* each and every backend is not overloaded 
* (optional) spillover happens along the ring (increasing hash)

Another test may issue a set of requests with different hash keys but again no backend should be overloaded.

Reply:
```go
func TestConsistentHashBoundedLoadDistribution(t *testing.T) {
	endpoints := []string{"http://127.0.0.1:8080", "http://127.0.0.2:8080", "http://127.0.0.3:8080"}
	r, _ := http.NewRequest("GET", "http://127.0.0.1:1234/foo", nil)
	route := NewAlgorithmProvider().Do([]*routing.Route{{
		Route: eskip.Route{
			BackendType: eskip.LBBackend,
			LBAlgorithm: ConsistentHash.String(),
			LBEndpoints: endpoints,
		},
	}})[0]
	ch := route.LBAlgorithm.(consistentHash)
	ctx := &routing.LBContext{Request: r, Route: route, Params: map[string]interface{}{ConsistentHashBalanceFactor: 1.25}}

	for i := 0; i < 100; i++ {
		ep := ch.Apply(ctx)
		ep.Metrics.IncInflightRequest() // done by proxy

		t.Logf("%d %d %d",
			route.LBEndpoints[ch[0].index].Metrics.GetInflightRequests(),
			route.LBEndpoints[ch[1].index].Metrics.GetInflightRequests(),
			route.LBEndpoints[ch[2].index].Metrics.GetInflightRequests())
	}
```
```
go test -run TestConsistentHashBoundedLoadDistribution ./loadbalancer -v
```
```
=== RUN   TestConsistentHashBoundedLoadDistribution
    algorithm_test.go:409: 1 0 0
    algorithm_test.go:409: 1 1 0
    algorithm_test.go:409: 2 1 0
    algorithm_test.go:409: 2 2 0
    algorithm_test.go:409: 3 2 0
    algorithm_test.go:409: 3 3 0
    algorithm_test.go:409: 3 3 1
    algorithm_test.go:409: 4 3 1
    algorithm_test.go:409: 4 4 1
    algorithm_test.go:409: 5 4 1
    algorithm_test.go:409: 5 5 1
    algorithm_test.go:409: 6 5 1
    algorithm_test.go:409: 6 6 1
    algorithm_test.go:409: 6 6 2
    algorithm_test.go:409: 7 6 2
    algorithm_test.go:409: 7 7 2
    algorithm_test.go:409: 8 7 2
    algorithm_test.go:409: 8 8 2
    algorithm_test.go:409: 8 8 3
    algorithm_test.go:409: 9 8 3
    algorithm_test.go:409: 9 9 3
    algorithm_test.go:409: 10 9 3
    algorithm_test.go:409: 10 10 3
    algorithm_test.go:409: 11 10 3
    algorithm_test.go:409: 11 11 3
    algorithm_test.go:409: 11 11 4
    algorithm_test.go:409: 12 11 4
    algorithm_test.go:409: 12 12 4
    algorithm_test.go:409: 13 12 4
    algorithm_test.go:409: 13 13 4
    algorithm_test.go:409: 13 13 5
    algorithm_test.go:409: 14 13 5
    algorithm_test.go:409: 14 14 5
    algorithm_test.go:409: 15 14 5
    algorithm_test.go:409: 15 15 5
    algorithm_test.go:409: 16 15 5
    algorithm_test.go:409: 16 16 5
    algorithm_test.go:409: 16 16 6
    algorithm_test.go:409: 17 16 6
    algorithm_test.go:409: 17 17 6
    algorithm_test.go:409: 18 17 6
    algorithm_test.go:409: 18 18 6
    algorithm_test.go:409: 18 18 7
    algorithm_test.go:409: 19 18 7
    algorithm_test.go:409: 19 19 7
    algorithm_test.go:409: 20 19 7
    algorithm_test.go:409: 20 20 7
    algorithm_test.go:409: 21 20 7
    algorithm_test.go:409: 21 21 7
    algorithm_test.go:409: 21 21 8
    algorithm_test.go:409: 22 21 8
    algorithm_test.go:409: 22 22 8
    algorithm_test.go:409: 23 22 8
    algorithm_test.go:409: 23 23 8
    algorithm_test.go:409: 23 23 9
    algorithm_test.go:409: 24 23 9
    algorithm_test.go:409: 24 24 9
    algorithm_test.go:409: 25 24 9
    algorithm_test.go:409: 25 25 9
    algorithm_test.go:409: 26 25 9
    algorithm_test.go:409: 26 26 9
    algorithm_test.go:409: 26 26 10
    algorithm_test.go:409: 27 26 10
    algorithm_test.go:409: 27 27 10
    algorithm_test.go:409: 28 27 10
    algorithm_test.go:409: 28 28 10
    algorithm_test.go:409: 28 28 11
    algorithm_test.go:409: 29 28 11
    algorithm_test.go:409: 29 29 11
    algorithm_test.go:409: 30 29 11
    algorithm_test.go:409: 30 30 11
    algorithm_test.go:409: 31 30 11
    algorithm_test.go:409: 31 31 11
    algorithm_test.go:409: 31 31 12
    algorithm_test.go:409: 32 31 12
    algorithm_test.go:409: 32 32 12
    algorithm_test.go:409: 33 32 12
    algorithm_test.go:409: 33 33 12
    algorithm_test.go:409: 33 33 13
    algorithm_test.go:409: 34 33 13
    algorithm_test.go:409: 34 34 13
    algorithm_test.go:409: 35 34 13
    algorithm_test.go:409: 35 35 13
    algorithm_test.go:409: 36 35 13
    algorithm_test.go:409: 36 36 13
    algorithm_test.go:409: 36 36 14
    algorithm_test.go:409: 37 36 14
    algorithm_test.go:409: 37 37 14
    algorithm_test.go:409: 38 37 14
    algorithm_test.go:409: 38 38 14
    algorithm_test.go:409: 38 38 15
    algorithm_test.go:409: 39 38 15
    algorithm_test.go:409: 39 39 15
    algorithm_test.go:409: 40 39 15
    algorithm_test.go:409: 40 40 15
    algorithm_test.go:409: 41 40 15
    algorithm_test.go:409: 41 41 15
    algorithm_test.go:409: 41 41 16
    algorithm_test.go:409: 42 41 16
    algorithm_test.go:409: 42 42 16
```

Reply:
> It could not happen that single backend gets +20 inflight requests when requests are balanced by the algorithm

Yes that can't happen, I just did it for the test so that it is obvious to the reader which one is overloaded / above average. I can add the test you show above though to show no server is overloaded as requests come in :+1: